cumulus/polkadot-omni-node/lib/src/common/runtime.rs

```diff
@@ -62,7 +101,109 @@ pub trait RuntimeResolver {
 pub struct DefaultRuntimeResolver;
 
 impl RuntimeResolver for DefaultRuntimeResolver {
-	fn runtime(&self, _chain_spec: &dyn ChainSpec) -> sc_cli::Result<Runtime> {
-		Ok(Runtime::Omni(BlockNumber::U32, Consensus::Aura(AuraConsensusId::Sr25519)))
+	fn runtime(&self, chain_spec: &dyn ChainSpec) -> sc_cli::Result<Runtime> {
+		let metadata_inspector = MetadataInspector::new(chain_spec)?;
+		let block_number = match metadata_inspector.block_number() {
+			Some(inner) => inner,
+			None => {
+				log::warn!(
+					r#"⚠️  There isn't a runtime type named `System`, corresponding to the `frame-system`
+                pallet (https://docs.rs/frame-system/latest/frame_system/). Please check Omni Node docs for runtime conventions:
+                https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/omni_node/index.html#runtime-conventions.
+                Note: We'll assume a block number size of `u32`."#
+				);
+				BlockNumber::U32
+			},
+		};
+
+		if !metadata_inspector.pallet_exists(DEFAULT_PARACHAIN_SYSTEM_PALLET_NAME) {
+			log::warn!(
+				r#"⚠️  The parachain system pallet (https://docs.rs/crate/cumulus-pallet-parachain-system/latest) is
+			   missing from the runtime’s metadata. Please check Omni Node docs for runtime conventions:
+			   https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_docs/reference_docs/omni_node/index.html#runtime-conventions."#
+			);
+		}
+
+		Ok(Runtime::Omni(block_number, Consensus::Aura(AuraConsensusId::Sr25519)))
+	}
+}
+
+/// Inspect runtime metadata.
+pub struct MetadataInspector(subxt::Metadata);
+
+impl MetadataInspector {
+	/// Creates an instance of `MetadataInspector`.
+	pub fn new(chain_spec: &dyn ChainSpec) -> Result<MetadataInspector, sc_cli::Error> {
+		MetadataInspector::get(chain_spec).map(MetadataInspector)
+	}
+
+	/// Checks if pallet exists in runtime's metadata based on pallet name.
+	pub fn pallet_exists(&self, name: &str) -> bool {
+		self.0.pallet_by_name(name).is_some()
+	}
+
+	/// Get the configured runtime's block number type from `frame-system` pallet storage.
+	pub fn block_number(&self) -> Option<BlockNumber> {
+		let pallet_metadata = self.0.pallet_by_name(DEFAULT_FRAME_SYSTEM_PALLET_NAME);
+		pallet_metadata
+			.and_then(|inner| inner.storage())
+			.and_then(|inner| inner.entry_by_name("Number"))
+			.and_then(|number_ty| match number_ty.entry_type() {
+				StorageEntryType::Plain(ty_id) => Some(ty_id),
+				_ => None,
+			})
+			.and_then(|ty_id| self.0.types().resolve(*ty_id))
+			.and_then(|portable_type| BlockNumber::from_type_def(&portable_type.type_def))
+	}
+
+	/// Get the runtime metadata from a wasm blob existing in a chain spec.
+	pub fn get(chain_spec: &dyn ChainSpec) -> Result<subxt::Metadata, sc_cli::Error> {
```

Comment:
I'd be more precise with the name here:
```suggestion
	pub fn fetch_metadata(chain_spec: &dyn ChainSpec) -> Result<subxt::Metadata, sc_cli::Error> {
```